src/commands/data/export/bulk.ts

```diff
@@ -0,0 +1,258 @@
+    const { flags } = await this.parse(DataExportBulk);
+
+    const conn = flags['target-org'].getConnection(flags['api-version']);
+
+    const timeout = flags.async ? Duration.minutes(0) : flags.wait ?? Duration.minutes(0);
+
+    // `flags['query-file']` will be present if `flags.query` isn't. oclif's `exclusive` isn't quite that clever
+    const soqlQuery = flags.query ?? fs.readFileSync(flags['query-file'] as string, 'utf8');
+
+    const lineEnding = flags['line-ending'] ?? platform() === 'win32' ? 'CRLF' : 'LF';
+
+    const columnDelimiter = flags['column-delimiter'] ?? 'COMMA';
+
+    const async = timeout.milliseconds === 0;
+
+    const baseUrl = flags['target-org'].getField<string>(Org.Fields.INSTANCE_URL).toString();
+
+    const ms = new MultiStageOutput<QueryJobInfoV2>({
+      jsonEnabled: flags.json ?? false,
+      stages: async
+        ? ['creating query job', 'done']
+        : ['creating query job', 'processing the job', 'exporting records'],
+      title: async ? 'Exporting data (async)' : 'Exporting data',
+      postStagesBlock: [
+        {
+          label: 'Status',
+          type: 'dynamic-key-value',
+          bold: true,
+          get: (data) => data?.state,
+        },
+        {
+          label: 'Job Id',
+          type: 'dynamic-key-value',
+          bold: true,
+          get: (data) =>
+            data?.id &&
+            terminalLink(
+              data.id,
+              `${baseUrl}/lightning/setup/AsyncApiJobStatus/page?address=${encodeURIComponent(`/${data.id}`)}`
+            ),
+        },
+      ],
+    });
+
+    // async: create query job in the org but don't poll for its status
+    if (async) {
+      const job = new QueryJobV2(conn, {
+        bodyParams: {
+          query: soqlQuery,
+          operation: flags['all-rows'] ? 'queryAll' : 'query',
+          columnDelimiter,
+          lineEnding,
+        },
+        pollingOptions: {
+          pollTimeout: timeout.milliseconds,
+          pollInterval: 5000,
+        },
+      });
+
+      job.on('error', (err) => {
+        ms.stop(err as Error);
+        throw err;
+      });
+
+      job.on('open', (jobInfo: QueryJobInfoV2) => {
+        ms.goto('done', { id: jobInfo.id });
+        ms.stop();
+      });
+
+      ms.goto('creating query job');
+
+      try {
+        const jobInfo = await job.open();
+
+        const cache = await BulkExportRequestCache.create();
+        await cache.createCacheEntryForRequest(
+          jobInfo.id,
+          {
+            filePath: flags['output-file'],
+            format: flags['result-format'],
+            columnDelimiter,
+          },
+          conn.getUsername(),
+          conn.getApiVersion()
+        );
+
+        this.log(messages.getMessage('export.timeout', [jobInfo.id]));
+
+        return {
+          jobId: jobInfo.id,
+          filePath: flags['output-file'],
+        };
+      } catch (err) {
+        const error = err as Error;
+        ms.stop(error);
+        throw err;
```

Comment:
Maybe an enhancement to `MultiStageOutput.stop()` would be to accept an unknown so this pattern wouldn't be all over the commands.  It's promoting an anti-pattern with `err as Error` coercion.